### PR TITLE
Add kernel item and nd_item support

### LINKED ISSUE #274: Support SYCL indexing function
- [ ] Analyze LLVM IR generated for SYCL indexing functions by dpcpp
- [ ] Design equivalent semantic in @dppy.kernel
- [ ] Implement support for SYCL indexing functions

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Add mocks for sycl's item and nd_item in python and add optional support for them as a first argument to experimental kernel. Changes does not introduce any changes to spriv kernel signature and works only at typing level.
So far there were provided two overloads:
- `Item.get_id(idx)`
- `NdItem.get_global_id(idx)`

Things that need to be done in following PRs:
- Add overloads for `get_local_id(idx)`
- Add `Group` type and overloads related to it
- Update Barrier syntax.
- Deprecate current way of getting indexes.

Checklist:
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

Fixes #274

Motivated by #1106

### CODE REVIEW

File: numba_dpex/experimental/__init__.py [hunk 1]
```diff
@@ -8,7 +8,10 @@
 
 from numba.core.imputils import Registry
 
-from ._kernel_dpcpp_spirv_overloads import _atomic_ref_overloads
+from ._kernel_dpcpp_spirv_overloads import (
+    _atomic_ref_overloads,
+    _item_overloads,
```
Comment (diptorupd) [inline numba_dpex/experimental/__init__.py:13]: Can we call the module `_index_space_id_overloads`?

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_item_overloads.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+@overload_method(
+    NdItemType, "get_global_id", target=DPEX_KERNEL_EXP_TARGET_NAME
+)
+def ol_nd_item_get_global_id(nd_item, dim):
+    """SPIR-V overload for
+    :meth:`numba_dpex.kernel_api.NdItem.get_global_id`.
+
+    Generates the same LLVM IR instruction as dpcpp for the
+    `sycl::item::get_global_id` function.
```
Comment (diptorupd) [inline numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_item_overloads.py:80]:
```suggestion
    `sycl::nd_item::get_global_id` function.
```

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_item_overloads.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    ty_context, ty_dim  # pylint: disable=unused-argument
+):
+    """Generates instruction for spirv i64 get_global_id(i32) call."""
+    sig = types.int64(types.int32)
+
+    def _intrinsic_exchange_gen(context, builder, sig, args):
+        [dim] = args
+        get_global_id = _declare_function(
+            # unsigned int - is what demangler returns from IR instruction
+            # generated by dpcpp. However int32 is passed as an argument.
```
Comment (diptorupd) [inline numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_item_overloads.py:33]: Actually, I think it is the reason Numba generates a lot of extra bound checks as we are passing a signed int to a function that expects unsigned int. See https://github.com/IntelPython/numba-dpex/issues/936. Filing it as a future TODO unrelated to current PR.

File: numba_dpex/experimental/core/types/kernel_api/items.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Defines numba types for Item and NdItem classes"""
+
+from numba.core import errors, types
+
+
+class ItemType(types.Type):
+    """Numba-dpex type corresponding to
```
Comment (diptorupd) [inline numba_dpex/experimental/core/types/kernel_api/items.py:11]: The docstring should fit under 80 chars on one line

File: numba_dpex/experimental/core/types/kernel_api/items.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    def key(self):
+        """Numba type specific overload"""
+        return self._ndim
+
+    def cast_python_value(self, args):
+        raise NotImplementedError
+
+
+class NdItemType(types.Type):
+    """Numba-dpex type corresponding to
```
Comment (diptorupd) [inline numba_dpex/experimental/core/types/kernel_api/items.py:38]: Docstring most likely will fit on one line < 80 chars

File: numba_dpex/experimental/kernel_dispatcher.py [hunk 6]
```diff
@@ -271,7 +271,7 @@ def _compile_cached(
                     "w",
                     encoding="UTF-8",
                 ) as fptr:
-                    fptr.write(cres.library.final_module)
+                    fptr.write(str(cres.library.final_module))
```
Comment (diptorupd) [inline numba_dpex/experimental/kernel_dispatcher.py:274]: :+1:

File: numba_dpex/experimental/launcher.py [hunk 7]
```diff
@@ -129,13 +135,28 @@ def _submit_kernel(  # pylint: disable=too-many-arguments
     else:
         sig = ty_return(ty_kernel_fn, ty_index_space, ty_kernel_args_tuple)
 
-    kernel_sig = types.void(*ty_kernel_args_tuple)
+    # Add Item/NdItem as a first argument to kernel arguments list. It is
+    # empty struct so any other modifications at kernel submission needed.
```
Comment (diptorupd) [inline numba_dpex/experimental/launcher.py:139]:
```suggestion
    # an empty struct so any other modifications at kernel submission are not needed.
```

File: numba_dpex/experimental/launcher.py [hunk 8]
```diff
@@ -129,13 +135,28 @@ def _submit_kernel(  # pylint: disable=too-many-arguments
...
+    ):
+        if isinstance(ty_index_space, RangeType):
+            ty_item = ItemType(ty_index_space.ndim)
+        else:
+            ty_item = NdItemType(ty_index_space.ndim)
+
+        ty_kernel_args_tuple = (ty_item, *ty_kernel_args_tuple)
+    else:
+        warnings.warn(
+            "Kernels without item/nd_item will be not supported in the future"
```
Comment (diptorupd) [inline numba_dpex/experimental/launcher.py:151]:
```suggestion
            "Kernels without item/nd_item will be not supported in the future",  DeprecationWarning
```

File: numba_dpex/kernel_api/__init__.py [hunk 9]
```diff
@@ -10,6 +10,13 @@
...
 from .atomic_ref import AtomicRef
+from .items import Item, NdItem
 from .memory_enums import AddressSpace, MemoryOrder, MemoryScope
 
-__all__ = ["AddressSpace", "AtomicRef", "MemoryOrder", "MemoryScope"]
+__all__ = [
+    "AddressSpace",
+    "AtomicRef",
+    "Item",
+    "NdItem" "MemoryOrder",
```
Comment (diptorupd) [inline numba_dpex/kernel_api/__init__.py:20]:
```suggestion
    "NdItem", "MemoryOrder",
```

File: numba_dpex/kernel_api/items.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
+# SPDX-FileCopyrightText: 2024 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Implements a mock Python classes to represent ``sycl::item`` and
```
Comment (diptorupd) [inline numba_dpex/kernel_api/items.py:5]: MY preference for the module name will be `index_space_ids.py`

File: numba_dpex/experimental/core/types/kernel_api/items.py (hunk 5 above)
Comment (ZzEeKkAa) [reply numba_dpex/experimental/core/types/kernel_api/items.py:38]: Yea, it got shrink since last rebase

**Review by diptorupd — APPROVED** [review]
Great work! Yet another 3 year old feature request can be marked as complete :smile:

**Pull request merged**